odoo/tools/safe_eval.py

```diff
@@ -1,225 +1,576 @@
-    :raises NameError: in case a forbidden name (containing two underscores)
-                       is found in ``code_obj``
+    Raises a :exc:`NameError` if the object is accessing a forbidden name.
 
-    .. note:: actually forbids every name containing 2 underscores
-    """
-    for name in code_obj.co_names:
-        if "__" in name or name in _UNSAFE_ATTRIBUTES:
-            raise NameError('Access to forbidden name %r (%r)' % (name, expr))
-
-def assert_valid_codeobj(allowed_codes, code_obj, expr):
-    """ Asserts that the provided code object validates against the bytecode
-    and name constraints.
-
-    Recursively validates the code objects stored in its co_consts in case
-    lambdas are being created/used (lambdas generate their own separated code
-    objects and don't live in the root one)
-
-    :param allowed_codes: list of permissible bytecode instructions
-    :type allowed_codes: set(int)
-    :param code_obj: code object to name-validate
-    :type code_obj: CodeType
-    :param str expr: expression corresponding to the code object, for debugging
-                     purposes
-    :raises ValueError: in case of forbidden bytecode in ``code_obj``
-    :raises NameError: in case a forbidden name (containing two underscores)
-                       is found in ``code_obj``
+    :param obj: is an object of any type that needs to be wrapped.
+    :param type_checker: is the function used to check the type of the object.
+                 The function has to follow the following signature: `type_checker(obj)`.
+                 You can refer to the :func:`__call_checker` function in this module for further documentation.
     """
-    assert_no_dunder_name(code_obj, expr)
 
-    # set operations are almost twice as fast as a manual iteration + condition
-    # when loading /web according to line_profiler
-    code_codes = {i.opcode for i in dis.get_instructions(code_obj)}
-    if not allowed_codes >= code_codes:
-        raise ValueError("forbidden opcode(s) in %r: %s" % (expr, ', '.join(opname[x] for x in (code_codes - allowed_codes))))
+    __slots__ = ["__obj", "__type_checker"]
 
-    for const in code_obj.co_consts:
-        if isinstance(const, CodeType):
-            assert_valid_codeobj(allowed_codes, const, 'lambda')
+    def __init__(self, obj, type_checker):
+        if isinstance(obj, _wrapped_instances) or (type(obj) is type and obj in _wrapped_types) or type(obj) is types.ModuleType and obj.__name__ in _ALLOWED_MODULES:  # noqa: E721 (Reason: `type(...) is ...` is stricter than `isinstance(..., ...)` in this case) # pylint: disable=unidiomatic-typecheck
+            self.__obj = obj
+        else:
+            self.__obj = type_checker(obj)
+
+        self.__type_checker = type_checker
+
+    def __getitem__(self, item):
+        self.__type_checker(item)
+        return self.__type_checker(self.__obj.__getitem__(item))
+
+    def __setitem__(self, key, value):
+        self.__type_checker(key)
```

Comment:
It's mostly used for lists

Reply:
same, list is in allowed_type, so it won't be encapsulated in SafeWrapper isn't it ?

```py
fn = lambda: [1, 2, 3]
safe_eval('print(fn())', {'fn': fn, 'print': print}
```

![image](https://github.com/odoo/odoo/assets/54316226/163faa26-1201-4c2b-8329-c5639cf527b6)

Reply:
> SafeWrapper is not used for standad dict, for which types is it useful so ? (__setitem__)

So that question is the same for list or dict, it's not used, so for which type is it useful ?

Reply:
```py
if isinstance(key, str) and "__" in key:
    raise NameError(f"Forbidden name '{key}'.")
```

but that part is still useless, because you can call the update method

Reply:
I'm working on that ;)